### PR TITLE
Imperial Magister, Chivalry, Prompts for career switches

### PR DESCRIPTION
**Pull request by Z3rca**
### Chivalry:
image

added the custom resource system for Bretonnia.

positive actions are providing chivalry, while negative will even take it away. Some actions like leveling up or clearing hideouts also provide chivalry. Letting people go (lords or do not imprision troops) also provide a bit chivalry. 

some knights cost chivalry on a daily basis and as upgrades (even early on)

the chivalry system is subsided in tiers. Depending on the amount of chivalry the player climbs certain levels. The levels increase or decrease the amount of morale and knight wages. Even though the effects are quite impactful on paper, the benefits and disadvantages feel at least after some hours of testing rather medicore, and rather more enganging for the early game of a campaign. 

Obviously, the numbers can be changed if there is need for adjustments.

### Prompts for career changes.

image

When the dialog options for vampire blood lines or new drinking from the black grail are about to trigger a career change, the player is kindly informed that this will happen, which is intended to avoid frustration 

Bretonnian knights can now talk to the serpent of mousillon to become a knight of mousillon.

Text etc. can be obviously changed. this is just an example. 

###  The Imperial College Wizard

image

The new (and last career for the empire for this update ) allows to become a full fledged college wizard. The clue is however not just a very diverse career covering for the current 6 (and later 8 ) winds, but a mini "crafting system" buffing units. With so called powerstones units can be magically enhanced. Each unit can have 1 upgrade.

Each wind has its own power stones, which makes more college wizards or a further subdivision obsolete. 

The crafting system is depended on a new Extended Party info. which stores the unit id, and an array of attributes(basically a dynamic attribute list). changing the stones changes also the attributes. Non existent units will be removed from the dictionary. Currently this dictionary stores such info for any party. It is potentially maybe better to have this restricted to the player.

Interestingly this mechanic could later also account for item enchantment. I keep it for now just for troops in order not overblow the whole thing in complexity. 

The array is put into the extended unit info deliberately.  I would like to extent on this in the future for other careers.

### CODE REVIEW

File: CSharpSourceCode/Extensions/ExtendedInfoSystem/MobilePartyExtendedInfo.cs [hunk 1]
```diff
@@ -11,5 +12,33 @@ public class MobilePartyExtendedInfo
     {
         [SaveableField(0)] public string CurrentBlessingStringId = null;
         [SaveableField(1)] public int CurrentBlessingRemainingDuration = -1;
+        [SaveableField(2)] public Dictionary<string, List<string>> TroopAttributes = new Dictionary<string, List<string>>();
+        
+        public void AddTroopAttribute(CharacterObject troop, string attribute)
+        {
+            if (TroopAttributes == null) TroopAttributes = new Dictionary<string, List<string>>();
+            if (!TroopAttributes.ContainsKey(troop.StringId))
```
Comment (hunharibo) [inline CSharpSourceCode/Extensions/ExtendedInfoSystem/MobilePartyExtendedInfo.cs]: use the TryGetValue pattern pls

File: CSharpSourceCode/BattleMechanics/StatusEffect/StatusEffectMissionLogic.cs [hunk 2]
```diff
@@ -31,7 +57,30 @@ public override void OnMissionTick(float dt)
                         comp.OnTick(dt);
                     }
                 }
+                if (!_unprocessedCharacters.Any()) continue;
+
+                if (_unprocessedCharacters.Contains(agent))
+                {
+                    CheckPermanentEffectsForAddingPermanentEffects(agent);
+                }
```
Comment (hunharibo) [inline CSharpSourceCode/BattleMechanics/StatusEffect/StatusEffectMissionLogic.cs]: use the Queue pattern instead. Better for performance, you are LINQ querying twice with Any then Contains

File: CSharpSourceCode/BattleMechanics/TORBattleAgentLogic.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 namespace TOR_Core.BattleMechanics
 {
+    //TODO this needs a renaming
```
Comment (hunharibo) [inline CSharpSourceCode/BattleMechanics/TORBattleAgentLogic.cs]: why?

File: CSharpSourceCode/BattleMechanics/TriggeredEffect/TriggeredEffect.cs [hunk 4]
```diff
@@ -109,7 +109,7 @@ public void Trigger(Vec3 position, Vec3 normal, Agent triggererAgent, AbilityTem
                     TORMissionHelper.ApplyStatusEffectToAgents(targets, effect.StringID, triggererAgent, statusEffectDuration, true, _isTemplateMutated);
                 }
             }
-
+            if(_template==null)return;
```
Comment (hunharibo) [inline CSharpSourceCode/BattleMechanics/TriggeredEffect/TriggeredEffect.cs]: why is this needed here? the entire code block starts with this very statement (line 39)

File: CSharpSourceCode/CampaignMechanics/CustomResources/CustomResourceManager.cs [hunk 5]
```diff
@@ -73,7 +89,126 @@ public static void RegisterEvents()
...
+                    customResourceGain.AddFactor(0.25f);
+                }
+                 
+                
+                Hero.MainHero.AddCultureSpecificCustomResource(customResourceGain.ResultNumber);
+
+            }
+        }
+
+        private void TournamentFinshedChange(CharacterObject winner, MBReadOnlyList<CharacterObject> participants,
```
Comment (hunharibo) [inline CSharpSourceCode/CampaignMechanics/CustomResources/CustomResourceManager.cs]: missing "i" from method name

File: CSharpSourceCode/Extensions/UI/PartyCharacterVMExtension.cs [hunk 6]
```diff
@@ -106,9 +105,12 @@ public void ExecuteButtonClick()
             SpecialbuttonEventManagerHandler.Instance.OnButtonClicked(troop,isPrisoner );
             
             ((PartyCharacterVM)_vm).RefreshValues();
+            RefreshValues();
```
Comment (hunharibo) [inline CSharpSourceCode/Extensions/UI/PartyCharacterVMExtension.cs]: is this really needed? the viewmodel extension should already call this after the base's

File: CSharpSourceCode/Extensions/UI/PartyViewExtension.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    [ViewModelExtension(typeof(PartyVM))]
+    public class PartyViewExtension : BaseViewModelExtension
+    {
+        public PartyViewExtension(ViewModel vm) : base(vm)
+        {
+
+            var partyVM = vm as PartyVM;
+            
+            
+            SpecialbuttonEventManagerHandler.Instance.SetCurrentPartyUiExtensionInstance(partyVM);
```
Comment (hunharibo) [inline CSharpSourceCode/Extensions/UI/PartyViewExtension.cs]: I am not a fan of this pattern. Why isn't the logic in this class? The object could easily get destroyed while the reference is still kept in the other class

File: CSharpSourceCode/Models/TORPartyMoraleModel.cs [hunk 8]
```diff
@@ -27,9 +28,37 @@ public override ExplainedNumber GetEffectivePartyMorale(MobileParty mobileParty,
             {
                 result.Add(TORPerks.SpellCraft.StoryTeller.SecondaryBonus, TORPerks.SpellCraft.StoryTeller.Name);
             }
-            if (mobileParty.HasBlessing("cult_of_lady"))
```
Comment (hunharibo) [inline CSharpSourceCode/Models/TORPartyMoraleModel.cs:30]: what was the blessing effect replaced with?

File: CSharpSourceCode/Utilities/TORConstants.cs [hunk 9]
```diff
@@ -29,5 +29,9 @@ public static class TORConstants
         public const int EMPIRE_VOICES_COUNT = 2;
         public const int BRETONNIA_VOICE_INDEX_START = 29;
         public const int BRETONNIA_VOICES_COUNT = 3;
+
+        public const string HERRIMAULT_CULTURE = "sea_raiders";
+        public const string BRETONNIA_CULTURE = "vlandia";
+        public const string SYLVANIA_CULTURE = "khuzait";
```
Comment (hunharibo) [inline CSharpSourceCode/Utilities/TORConstants.cs:36]: This is a great idea, but even though you added these, you continue to hardcode the culture strings all throughout. If we were to do this, we would have to find and replace all culture strings in the entire codebase, it's not worth it otherwise.

File: ModuleData/tor_strings.xml [hunk 10]
```diff
@@ -852,6 +895,7 @@
   <string id="tor_damagetype.Magical" text="{=tor_damagetype_magical_str}Magical"/>
   <string id="tor_damagetype.Fire" text="{=tor_damagetype_fire_str}Fire"/>
   <string id="tor_damagetype.Holy" text="{=tor_damagetype_holy_str}Holy"/>
+  <string id="tor_damagetype.Frost" text="{=tor_damagetype_holy_str}Holy"/>
```
Comment (hunharibo) [inline ModuleData/tor_strings.xml]: text is probably unintended to be holy?

File: CSharpSourceCode/BattleMechanics/TORBattleAgentLogic.cs (hunk 3 above)
Comment (Z3rca) [reply CSharpSourceCode/BattleMechanics/TORBattleAgentLogic.cs]: I guess it's okay, i was just confused that the whole logic seems just to be summoning related.

File: CSharpSourceCode/BattleMechanics/TriggeredEffect/TriggeredEffect.cs (hunk 4 above)
Comment (Z3rca) [reply CSharpSourceCode/BattleMechanics/TriggeredEffect/TriggeredEffect.cs]: its a debug problem. when i am debugging the template sometimes gets null, I often add that to avoid unnecessary crashs during debugging. I'll remove it

File: CSharpSourceCode/Models/TORPartyMoraleModel.cs (hunk 8 above)
Comment (Z3rca) [reply CSharpSourceCode/Models/TORPartyMoraleModel.cs:30]: you gain now chivalry.

File: CSharpSourceCode/Utilities/TORConstants.cs (hunk 9 above)
Comment (Z3rca) [reply CSharpSourceCode/Utilities/TORConstants.cs:36]: lets say i first wanted to have your feedback before I touch 29 files :D

Comment (Z3rca) [reply CSharpSourceCode/Utilities/TORConstants.cs:36]: done

File: CSharpSourceCode/Extensions/ExtendedInfoSystem/MobilePartyExtendedInfo.cs (hunk 1 above)
Comment (Z3rca) [reply CSharpSourceCode/Extensions/ExtendedInfoSystem/MobilePartyExtendedInfo.cs]: hope its now better

File: CSharpSourceCode/BattleMechanics/StatusEffect/StatusEffectMissionLogic.cs (hunk 2 above)
Comment (Z3rca) [reply CSharpSourceCode/BattleMechanics/StatusEffect/StatusEffectMissionLogic.cs]: done

File: CSharpSourceCode/CampaignMechanics/CustomResources/CustomResourceManager.cs (hunk 5 above)
Comment (Z3rca) [reply CSharpSourceCode/CampaignMechanics/CustomResources/CustomResourceManager.cs]: fixed

File: CSharpSourceCode/Extensions/UI/PartyViewExtension.cs (hunk 7 above)
Comment (Z3rca) [reply CSharpSourceCode/Extensions/UI/PartyViewExtension.cs]: please have a look now. The view model is provided by the extension, and gets removed whenever the actual instance gets removed.

**Review by hunharibo — APPROVED** [review]
